### PR TITLE
Document Functions, Combination Function, Troubleshooting guide

### LINKED ISSUE #795: Document Functions in Labels, combination functions etc
Document;

- [ ] what functions are,
- [ ] how functions are calculated 
- [ ] Provide examples
- [ ] #803

### PR DESCRIPTION
**Pull request by whitingjr**
## Fixes Issue

fixes #795 

## Changes proposed

[x] - what Functions are
[x] - Provide examples
[x] - #803  

## Check List (Check all the applicable boxes) 

- [ ] My change requires changes to the documentation.

### CODE REVIEW

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        rssFirstRequest: rssFirstRequest[i],
+        maxThroughputDensity: maxThroughputDensity[i],
+        buildId: buildId,
+        buildUrl: buildUrl,
+        version: ((name.split('-')[0].substring(0, 6) == 'spring' ) ? springVersion: quarkusVersion )
+    }))
+    return map;
+}
+``` 
+
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tasks/define-functions/index.md:77]: Can we have abridged json documents showing input & output from this function?

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+## Example Combination Functions
+
+`Combination Functions` rely on a returned value that is an Object, Array or individual value.
+This `Function` will conditionally reduce an array of values unless there is only a single value of type number.
+
+```javascript
+value => typeof value === "number" ? value : value.reduce((a, b) => Math.max(a, b))
+```
+
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tasks/define-functions/index.md:86]: see above for json example

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+This `Function` will create a value of the amount of time per request with the exponent rounded to 2 figures.
+
+```javascript
+value => (value.duration / value.requests).toFixed(2)
+```
+
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tasks/define-functions/index.md:92]: see above for json example

File: docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+* The `typeof` function to discover the data-type of the `Extractor` variable, object property or array element in the Label `Combination Function`, using the `Test label calculation` button to reveal the Javascript data-type
+{{% imgproc use-typeof-function Fit "785x769" %}}
+Test label calculation
+{{% /imgproc %}}
+* View the `Labels Log` and display DEBUG log level, reveals syntax errors when the function is executed
+* Use the `Show label values` button, provides a quick visual check of resulting values for the `Run` `Dataset`
+{{% imgproc buttons-on-dataset-view Fit "1091x462 " %}}
+Run Dataset view
+{{% /imgproc %}}
+Effective label values reveal the anticipated value or error text which can be due to the following:
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md:23]: Can we show a screenshot of `Labels Log` as well, there is debug information displayed showing the input values and the JS evaluation error message.  This is vital for users to debug problems when a function evaluation fails

**Review by johnaohara — CHANGES_REQUESTED** [review]
@whitingjr please see comments, thanks

File: docs/site/content/en/docs/Concepts/core-concepts/index.md [hunk 5]
```diff
@@ -88,6 +88,9 @@ It is possible for a [`Run`](#run) to include multiple [`Datasets`](#dataset), a
 
 A `Fingerprint` is combination of [`Filtering labels`](#label) that unique identifies comparable [`datasets`](#dataset) within a [`test`](#test)
 
+## Function
+A `Function` is a Javascript function that is executed on the server side. Functions can be used for validating expected data formats and substitution. Also used in `Combination Function`s to create derived metrics.
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Concepts/core-concepts/index.md:92]: sure, good point

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+ Using `Functions` in `Horreum` is a feature that provides a great deal of bespoke functionality to `Horreum` that is under the control of a user. The ability to use a `Function` written in Javascript makes it possible in the web User Interface (UI) to have:
+
+These Functions can be categorized as:
+
+* Selector (filter)
+* Transforming
+* Combination
+
+ When using `Horreum` you will find in `Functions` used in these Horreum objects:
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:21]: I'll remove

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+> 2. [uploaded](/docs/tasks/upload-run/) some data
+
+ Using `Functions` in `Horreum` is a feature that provides a great deal of bespoke functionality to `Horreum` that is under the control of a user. The ability to use a `Function` written in Javascript makes it possible in the web User Interface (UI) to have:
+
+These Functions can be categorized as:
+
+* Selector (filter)
+* Transforming
+* Combination
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:19]: sure, I can add that.

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+---
+> **Prerequisites**: You have already
+
+> 1. [created a Test](/docs/tasks/create-new-test/)
+
+> 2. [uploaded](/docs/tasks/upload-run/) some data
+
+ Using `Functions` in `Horreum` is a feature that provides a great deal of bespoke functionality to `Horreum` that is under the control of a user. The ability to use a `Function` written in Javascript makes it possible in the web User Interface (UI) to have:
+
+These Functions can be categorized as:
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:15]: will rework this section.

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+* Category formatter (Report) Function
+* Series (Report) Function
+* Series formatter (Report) Function
+* Scale Label (Report) Function
+* Scale Label formatter (Report) Function
+* Component Transformation (Report) Function
+
+## Making use of Horreum Functions
+
+ Javascript [ECMAScript 2023](https://262.ecma-international.org/14.0/) specification is available throughout Horreum `Functions`. A reference for Javascript functions based use is available [here](https://www.w3schools.com/jsref/default.asp)
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:42]: ok

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+## Example Filtering Functions
+
+These Functions rely on a condition evaluation to return a boolean value.
+The following will filter based on the individual Label Extractor only having the value 72.
+
+```javascript
+value => value === 73
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:50]: don't know where this came from. will fix that.

File: docs/site/content/en/docs/Tasks/define-functions/index.md (hunk 1 above)
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:77]: sure

File: docs/site/content/en/docs/Tasks/define-functions/index.md (hunk 3 above)
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:92]: sure

File: docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md (hunk 4 above)
Comment (whitingjr) [reply docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md:23]: Sure. I will add this.

**Review by whitingjr — COMMENTED** [review]
Changes have been done.

File: docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md
Comment (johnaohara) [inline docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md:1]: tutorials should have a defined goal for users to learn something, including step by step instructions and a defined outcome.  I don't thin this is a tutorial, but is a Reference Guide

File: docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
+---
+title: Troubleshooting Functions
+description: Steps to troubleshoot use of Functions in Horreum
+date: 2024-02-27
+weight: 4
+---
+
+## Troubleshooting Functions and Combination Function
+
+Creating Scheam Labels is one of the key capabilities that enables working with loosely defined data. Using a Validation Schema provides useful guarentees data in an uploaded Run JSON conforms to expected data types. Validation is optional and when not used maks the handling of metrics unreliable as data types may change.
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md:10]:
```suggestion
Creating Schema Labels is one of the key capabilities that enables working with loosely defined data. Using a Validation Schema provides useful guarantees data in an uploaded Run JSON conforms to expected data types. Validation is optional and when not used makes the handling of metrics unreliable as data types may change.  
```

File: docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+description: Steps to troubleshoot use of Functions in Horreum
+date: 2024-02-27
+weight: 4
+---
+
+## Troubleshooting Functions and Combination Function
+
+Creating Scheam Labels is one of the key capabilities that enables working with loosely defined data. Using a Validation Schema provides useful guarentees data in an uploaded Run JSON conforms to expected data types. Validation is optional and when not used maks the handling of metrics unreliable as data types may change.  
+
+To troubleshoot use of Javascript Functions the following techniques can be used
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md:12]:
```suggestion
To troubleshoot use of JavaScript Functions the following techniques can be used
```

File: docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+* The `typeof` function to discover the data-type of the `Extractor` variable, object property or array element in the Label `Combination Function`, using the `Test label calculation` button to reveal the Javascript data-type
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md:14]:
```suggestion
* The `typeof` function to discover the data-type of the `Extractor` variable, object property or array element in the Label `Combination Function`, using the `Test label calculation` button to reveal the JavaScript data-type
```

File: docs/site/content/en/docs/Concepts/core-concepts/index.md [hunk 14]
```diff
@@ -49,7 +49,7 @@ An `Extractor` is a **required** JSONPath expression that refers to a section of
 {{% alert title="Note" %}}In the majority of cases, an Extractor will simply point to a single, scalar value{{% /alert %}}
 
 ## Combination Function:
-A Combination Function is an optional Javascript function that takes all Extractor values as input and produces a Label value. 
+A Combination Function is an optional Javascript function that takes all Extractor values as input and produces a Label value. See [Function](#function)
```
Comment (johnaohara) [inline docs/site/content/en/docs/Concepts/core-concepts/index.md:52]:
```suggestion
A Combination Function is an optional JavaScript function that takes all Extractor values as input and produces a Label value. See [Function](#function)
```

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+date: 2024-02-29
+weight: 4
+---
+> **Prerequisites**: You have already
+
+> 1. [created a Test](/docs/tasks/create-new-test/)
+
+> 2. [uploaded](/docs/tasks/upload-run/) some data
+
+ Using `Functions` in `Horreum` is a feature that provides a great deal of bespoke functionality to `Horreum` that is under the control of a user. The ability to use a `Function` written in Javascript.
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tasks/define-functions/index.md:13]:
```suggestion
 Using `Functions` in `Horreum` is a feature that provides a great deal of bespoke functionality to `Horreum` that is under the control of a user. The ability to use a `Function` written in JavaScript.
```

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+* Category formatter (Report) Function
+* Series (Report) Function
+* Series formatter (Report) Function
+* Scale Label (Report) Function
+* Scale Label formatter (Report) Function
+* Component Transformation (Report) Function
+
+## Making use of Horreum Functions
+
+ Javascript [ECMAScript 2023](https://262.ecma-international.org/14.0/) specification is available throughout Horreum `Functions`.
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tasks/define-functions/index.md:43]:
```suggestion
 JavaScript [ECMAScript 2023](https://262.ecma-international.org/14.0/) specification is available throughout Horreum `Functions`.
```

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+
+## Example Filtering Functions
+
+These Functions rely on a condition evaluation to return a boolean value.
+The following will filter based on the individual Label Extractor only having the value 72.
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tasks/define-functions/index.md:48]:
```suggestion
The following will filter based on the individual `Label Extractor` only having the value 72.
```

File: docs/site/content/en/docs/Tasks/define-schema-and-views/index.md [hunk 18]
```diff
@@ -31,9 +31,12 @@ Switch tab to 'Labels' and add two **labels**: let's call first `ci-url` where y
...
+## Combination Functions
 
-Note that this function is calculated server-side and it is cached it the database. It is automatically recalculated when the label definition changes; the datasets are processed asynchronously, though, so there might be some lag.
+In the situation a derived metric is necessary the `Combination Function` is used to calculate the value. 
+Here we create a seperate Label called `throughput` and we extract two elements from the JSON document: `requests` using `$.results.requests` and `duration` using `$.results.duration`. In this case the input to the calculation function will be an object with two fields: `requests` and `duration`. You can get throughput by adding combination function (in Javascript) in the lambda syntax `value => value.requests / value.duration`. Had we used only one JSON Path the only argument to the function would be directly the result of the JSON Path query.
```
Comment (johnaohara) [inline docs/site/content/en/docs/Tasks/define-schema-and-views/index.md:37]:
```suggestion
Here we create a seperate Label called `throughput` and we extract two elements from the JSON document: `requests` using `$.results.requests` and `duration` using `$.results.duration`. In this case the input to the calculation function will be an object with two fields: `requests` and `duration`. You can get throughput by adding combination function (in JavaScript) in the lambda syntax `value => value.requests / value.duration`. Had we used only one JSON Path the only argument to the function would be directly the result of the JSON Path query.
```

**Review by johnaohara — CHANGES_REQUESTED** [review]

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+        buildId: buildId,
+        buildUrl: buildUrl,
+        version: ((name.split('-')[0].substring(0, 6) == 'spring' ) ? springVersion: quarkusVersion )
+    }))
+    return map;
+}
+```
+```json
+{
+  "runtime": "spring", "buildType": "native", "rssStartup": 55, "maxRss": 15, "avBuildTime": 1234, "avTimeToFirstRequest": 5, "avThroughput": 25, "rssFirstRequest": 5000, "maxThroughputDensity": 15, "buildId": "x512", "buildUrl": "http://acme.com", "version": "3.0"
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:96]: Sure.

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
+```javascript
+value => value === 72
+```
+
+## Example Transforming Functions
+
+`Transforming Functions` rely on a returned value that is an Object, Array or individual value.
+This `Function` relies on 12 `Extractors` setup on the `Schema Label`. Each Extractor configured to obtain an array of data items (except buildId and buildUrl).
+
+```json
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:59]: good idea

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 21]
```diff
@@ -0,0 +1,135 @@
...
+        avTimeToFirstRequest: avTimeToFirstRequest[i],
+        avThroughput: avThroughput[i],
+        rssFirstRequest: rssFirstRequest[i],
+        maxThroughputDensity: maxThroughputDensity[i],
+        buildId: buildId,
+        buildUrl: buildUrl,
+        version: ((name.split('-')[0].substring(0, 6) == 'spring' ) ? springVersion: quarkusVersion )
+    }))
+    return map;
+}
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:92]: sure

File: docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md
Comment (whitingjr) [reply docs/site/content/en/docs/Tutorials/troubleshooting-functions/index.md:1]: Good point, I'll relocate this.

File: docs/site/content/en/docs/Tasks/define-functions/index.md [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+* Change Detection (Test) Variable Calculation Function
+* View (Test) Component Rendering Function
+* Filtering (Report) Function
+* Category (Report) Function
+* Category formatter (Report) Function
+* Series (Report) Function
+* Series formatter (Report) Function
+* Scale Label (Report) Function
+* Scale Label formatter (Report) Function
+* Component Transformation (Report) Function
```
[the remark this answers is not in the document]
Comment (whitingjr) [reply docs/site/content/en/docs/Tasks/define-functions/index.md:39]: ok, I'll simplify and and add links where relevant.

**Comment by whitingjr** [comment]
Changes have been completed.

**Review by johnaohara — APPROVED** [review]
LGTM

**Comment by johnaohara** [comment]
@whitingjr please can you squash commits

**Comment by whitingjr** [comment]
@johnaohara squashed

**Pull request merged**